tensorflow/c/c_api_test.cc

```diff
@@ -1534,11 +1534,13 @@ TEST(CAPI, TestBitcastFrom_Reshape) {
   EXPECT_EQ(6 * TF_DataTypeSize(TF_UINT64), TF_TensorByteSize(a));
   EXPECT_EQ(6 * TF_DataTypeSize(TF_UINT64), TF_TensorByteSize(b));
 
-  // Check that a write to one tensor shows up in the other.
-  *(static_cast<int64_t*>(TF_TensorData(a))) = 4;
-  EXPECT_EQ(4, *(static_cast<int64_t*>(TF_TensorData(b))));
-  *(static_cast<int64_t*>(TF_TensorData(b))) = 6;
-  EXPECT_EQ(6, *(static_cast<int64_t*>(TF_TensorData(a))));
+  if (tensorflow::port::kLittleEndian) {
```

Comment:
Hi @cantonios ,

Thanks for the comment. The reason why it does not apply on BE machines is that the output Tensor of Bitcast op does not share a common buffer with the input Tensor, so the write to one tensor does not show up in the other.

Reply:
I understand.  See the other comment.  I think this is a behavior we likely want to preserve.  Especially in this test, where the two bit-casted types are both uint64_t.

Reply:
Well, I'm thinking we likely don't want a deep copy at all, and instead fail if you try to bit-cast something weird like 10 floats to 20 bfloat16s.

std::bit_cast only makes sense when bit-casting between types of the same size.  The only other use-case I'm aware of is for inspecting individual bytes, by bit-casting to a byte tensor (e.g. uint8_t).  *Maybe* there are other valid use-cases, but I don't know of them, and they're probably doing something fishy.  It would be good to know if there are any such cases currently in TF.

Reply:
@cantonios Thanks for the explanation! I think test cases which this PR aims to fix are doing bit-casting between types of different sizes.

For example, `//tensorflow/compiler/tests:cast_ops_test_cpu` is bit-casting from `float16` to `float32`. If we do not apply byte swapping on BE machines, these test cases will fail. And if we use byte swapping without a deep copy, the contents of the input Tensor would be changed after swapping.

Reply:
Hi @cantonios , I still have one question. In test case `//tensorflow/compiler/tests:cast_ops_test_cpu`, I think it's comparing the results between the JIT Compiled XLA Bitcast operation with the basic tensor Bitcast operation. 

https://github.com/tensorflow/tensorflow/blob/6cd264ba65e6161164d7f1b463b7d2415ac0aeb0/tensorflow/compiler/tests/cast_ops_test.py#L35-L47
 
The JIT Compiled XLA Bitcast operation (`compiled_f()`) will give the result in LE format on BE machines, so I guess it has used a copy between input and output buffers, otherwise after the execution of `compiled_f(x)` which involves byte swapping, the data in `x` would've been changed. But through debugging I found the data in `x` wasn't changed. 

If we do not apply deep copy and byte swapping in the basic tensor Bitcast operation, `f(x)` will give BE specific result, and the test will fail on s390x.  

Should we keep the behaviours of different Bitcast implementations consistent on BE machines? Please let me know what you think. Thanks again!